Implement self-conditioned CTC

komatta-san opened the pull request:
This implements self-conditioned CTC for Transformer and Conformer encoders.
https://arxiv.org/abs/2104.02724

sw005320 commented:
This https://github.com/espnet/espnet/runs/4467881330?check_suite_focus=true#step:8:36 does not seem to be related to this PR.
@Emrys365, do you have any idea?

Emrys365 commented:
I think it is related to the new positional argument `odim` added to the transformer encoder: https://github.com/espnet/espnet/blob/3b96776cb75afcdc3f7ae0ff0f227ce156f88271/espnet/nets/pytorch_backend/transformer/encoder.py#L84-L88

The TransformerSeparator is dependent on the base transformer implementation. So it should also be modified to match the new interface.

sw005320 commented:
Oh, yes.
You're right.
@komatta-san, can you fix it?

komatta-san commented:
I modified `odim` to be a named argument and moved it to the last position.

pengchengguo commented on espnet/nets/pytorch_backend/conformer/encoder.py:
```diff
@@ -93,6 +93,8 @@ def __init__(
         stochastic_depth_rate=0.0,
         intermediate_layers=None,
+        ctc_softmax=None,
+        odim=None,
```
I prefer to use `conditioning_layer_dim` instead of `odim` here. Although the output dimension of CTC is vocabulary size (`odim`), using another name like `conditioning_layer_dim` may reduce ambiguity. It is okay to keep `odim`.

komatta-san replied:
I will change it to `conditioning_layer_dim`.

pengchengguo reviewed (commented):
Thanks, @komatta-san.
Most of the codes are good to me except for some minor suggestions. Besides, please add some CI tests for the new components.

komatta-san commented on espnet/nets/pytorch_backend/conformer/encoder.py:
```diff
@@ -261,10 +267,22 @@ def forward(self, xs, masks):
+                        intermediate_result = self.ctc_softmax(encoder_output)
+
+                        if isinstance(xs, tuple):
+                            x, pos_emb = xs[0], xs[1]
```
I was trying to write it simply as well. But tuple `xs[0]` is immutable, so it can not be written like your suggestion.

komatta-san commented:
@pengchengguo 
Thank you for your suggestion. I modified the name of `odim` to `conditioning_layer_dim`.
For tests, I am not sure how to write it correctly, but I have added lines to  `test/test_e2e_asr_{transformer,conformer}.py`.

sw005320 commented:
Thanks, @komatta-san!
